File: src/Samples/XAML_Features/Validation/Person.cs
using System;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;

namespace OpenSilver.Samples.Showcase
{
    public class Person : INotifyPropertyChanged
    {
        private string _name;
        public string Name
        {
            get { return _name; }
            set
            {
                if (string.IsNullOrWhiteSpace(value))
                {
                    throw new Exception("Name cannot be empty.");
                }
                _name = value;
                RaisePropertyChanged("Name");
            }
        }

        private int _age = 1;
        [Display(Name = "Age", Description = "Age as positive number")]
        public int Age
        {
            get { return _age; }
            set
            {
                if (value <= 0)
                {
                    throw new Exception("Age cannot be lower than 0.");
                }
                _age = value;
                RaisePropertyChanged("Age");
            }
        }

        private string _email;
        [RegularExpression(@"^([a-zA-Z0-9\!\%\$\%\*\/\?\|\^\{\}\`\~\&\'\+\-\=_]\.?)*[a-zA-Z0-9\!\%\$\%\*\/\?\|\^\{\}\`\~\&\'\+\-\=_]@((([a-zA-Z0-9\!\%\$\%\*\/\?\|\^\{\}\`\~\&\'\+\-\=_]\.?)*[a-zA-Z0-9\!\%\$\%\*\/\?\|\^\{\}\`\~\&\'\+\-\=_])|(\[\d+\.\d+\.\d+\.\d+\]))$", ErrorMessage = "Not a valid e-mail address.")]
        [Display(Name = "Email", Description = "An e-mail address of the form <name>@<domain>, such as john@johndoe.com")]
        [Required]
        public string Email
        {
            get => _email;
            set
            {
                if (value != _email)
                {
                    Validator.ValidateProperty(value, new ValidationContext(this, null, null) { MemberName = "Email" });
                    _email = value;
                    RaisePropertyChanged(nameof(Email));
                }
            }
        }

        private void RaisePropertyChanged(string propertyName)
        {
            PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }

        public event PropertyChangedEventHandler PropertyChanged;
    }
}

File: vbsrc/Samples/XAML_Features/Validation/Person.vb
Imports System
Imports System.ComponentModel

#If SLMIGRATION
#Else
using Windows.Foundation;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Controls.Primitives;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Input;
using Windows.UI.Xaml.Media;
using Windows.UI.Xaml.Navigation;
#End If

Namespace Global.OpenSilver.Samples.VBShowcase
    'Validation:
    Public Class Person
        Implements INotifyPropertyChanged
        Private _name As String
        Public Property Name As String
            Get
                Return _name
            End Get
            Set(ByVal value As String)
                If String.IsNullOrWhiteSpace(value) Then
                    Throw New Exception("Name cannot be empty.")
                End If
                _name = value
                RaisePropertyChanged("Name")
            End Set
        End Property

        Private _age As Integer = 1
        Public Property Age As Integer
            Get
                Return _age
            End Get
            Set(ByVal value As Integer)
                If value <= 0 Then
                    Throw New Exception("Age cannot be lower than 0.")
                End If
                _age = value
                RaisePropertyChanged("Age")
            End Set
        End Property

        Private Sub RaisePropertyChanged(ByVal propertyName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
        End Sub
        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
    End Class
End Namespace
